tanjun/dependencies/reloader.py

```diff
@@ -0,0 +1,388 @@
+        self.sys_path = sys_path
+        self.last_modified_at = last_modified_at
+
+
+def _scan_one(path: pathlib.Path, /) -> typing.Optional[int]:
+    try:
+        return path.stat().st_mtime_ns
+
+    except FileNotFoundError:  # TODO: catch other errors here like perm errors
+        return None
+
+
+class _ScanResult:
+    __slots__ = ("py_paths", "removed_py_paths", "removed_sys_paths", "sys_paths")
+
+    def __init__(self) -> None:
+        self.py_paths: dict[str, _PyPathScanInfo] = {}
+        self.removed_py_paths: list[str] = []
+        self.removed_sys_paths: list[pathlib.Path] = []
+        self.sys_paths: dict[pathlib.Path, int] = {}
+
+
+class Reloader:
+    __slots__ = (
+        "_dead_unloads",
+        "_directories",
+        "_interval",
+        "_py_paths",
+        "_py_path_to_sys_path",
+        "_sys_paths",
+        "_task",
+        "_unload_on_delete",
+        "_waiting_for_py",
+        "_waiting_for_sys",
+    )
+
+    def __init__(
+        self,
+        *,
+        interval: typing.Union[int, float, datetime.timedelta] = datetime.timedelta(microseconds=500000),
+        unload_on_delete: bool = True,
+    ) -> None:
+        if isinstance(interval, datetime.timedelta):
+            interval = interval.total_seconds()
+
+        else:
+            interval = float(interval)
+
+        self._dead_unloads: set[typing.Union[str, pathlib.Path]] = set()
+        """Set of modules which cannot be unloaded."""
+
+        self._directories: dict[pathlib.Path, _DirectoryEntry] = {}
+        """Dict of paths to the metadata of tracked directories."""
+
+        self._interval = interval
+        """How often this should scan for changes and (re)load modules."""
+
+        self._py_paths: dict[str, _PyPathInfo] = {}
+        """Dict of module paths to info of the modules being targeted."""
+
+        self._sys_paths: dict[pathlib.Path, typing.Optional[int]] = {}
+        """Dict of system paths to info of files being targeted."""
+
+        self._task: typing.Optional[asyncio.Task[None]] = None
+        self._unload_on_delete = unload_on_delete
+
+        self._waiting_for_py: dict[str, int] = {}
+        """Dict of module paths to the new edit time of a module that's scheduled for a reload."""
+
+        self._waiting_for_sys: dict[pathlib.Path, int] = {}
+        """Dict of system paths to the new edit time of a module that's scheduled for a reload."""
+
+    def add_to_client(self: _ReloaderT, client: tanjun.Client, /) -> _ReloaderT:
+        (
+            client.add_client_callback(tanjun.ClientCallbackNames.STARTED, self.start)
+            .add_client_callback(tanjun.ClientCallbackNames.CLOSING, self.stop)
+            .set_type_dependency(Reloader, self)
+        )
+        if client.is_alive and client.loop:
+            client.loop.call_soon_threadsafe(self.start, client)
+
+        return self
+
+    async def add_modules_async(self: _ReloaderT, *paths: typing.Union[str, pathlib.Path]) -> _ReloaderT:
```

Comment:
should this and the other relevant methods take `(path, /, *other_paths)` to avoid accidentally calling them with 0 paths